nvm.sh

```diff
@@ -1851,7 +1851,7 @@ nvm_get_arch() {
   esac
 
   # If running a 64bit ARM kernel but a 32bit ARM userland, change ARCH to 32bit ARM (armv7l)
-  L=$(ls -dl /sbin/init 2>/dev/null) # if /sbin/init is 32bit executable
+  [ -f /sbin/init ] && L=$(ls -dl /sbin/init 2>/dev/null) # if /sbin/init is 32bit executable
```

Comment:
i believe this means `L` won't be unconditionally set, which means the next line could error if the `u` option is set. Can we move the condition inside the subshell?

Reply:
Makes sense – did that, thanks!

Reply:
Makes sense – did that, thanks!

Reply:
As to the `-e` question / the more global issue: I have not set it in my environment (and the issue also occurs in a clean chroot without any custom stuff). Still, there *is* an `errexit` set in the script that is used by Arch for building a package (https://gitlab.archlinux.org/pacman/pacman/-/blob/master/scripts/makepkg.sh.in#L387-406) – but that's been there for a while (while it has only become an issue with `nvm` sometime during the past few weeks, I think).

I'll fully admit that my bash / shell scripting knowledge is way to limited to properly assess if that might still be the root cause and/or how to address it there instead (or find out what other change over there might've caused it). `nvm` was the easier target for me ;)

Reply:
It's getting even more, hm, interesting?

With the check in the subshell, it does fail again as it did with the original code – which I didn't expect!

With a minimal example, like:

```bash
#!/bin/bash

set -e

[ -f /sbin/somethingnonexisting ] && L=$(ls -dl /sbin/somethingnonexisting 2>/dev/null)
echo '1'
L=$([ -f /sbin/somethingnonexisting ] && ls -dl /sbin/somethingnonexisting 2>/dev/null)
echo '2'
```

the resulting output would only be
```
1
```

This could be worked around by forcing it to return something "successful", but that doesn't seem like good/clean code to me.

```bash
#!/bin/bash

set -e

[ -f /sbin/somethingnonexisting ] && L=$(ls -dl /sbin/somethingnonexisting 2>/dev/null)
echo '1'
L=$([ -f /sbin/somethingnonexisting ] && ls -dl /sbin/somethingnonexisting 2>/dev/null || true)
echo '2'
```

Interestingly, it seems like we can get around the whole issue by just using `local` (we are in a function here in the `nvm.sh`, so that should be fine) – came upon this at https://unix.stackexchange.com/questions/23026/how-can-i-get-bash-to-exit-on-backtick-failure-in-a-similar-way-to-pipefail/88338#88338.

This seems to work with bash, zsh and dash, so I'd assume it should work "everywhere"?

A simple standalone example:

```bash
#!/bin/bsh

set -e

whatever(){
    [ -f /sbin/somethingnonexisting ] && L=$(ls -dl /sbin/somethingnonexisting 2>/dev/null)
    echo '1'
    local L=$(ls -dl /sbin/somethingnonexisting 2>/dev/null)
    echo '2'
}
```

This doesn't bail out early and gives the output:
```
1
2
```

as hoped for.

To sum it all up:

`local L=$(ls -dl /sbin/init 2>/dev/null) # if /sbin/init is 32bit executable` should be all we'd need to change this line to, if I'm not mistaken – would that sound good to you?

Reply:
isn't that what the line already is?

Reply:
Not entirely – but almost. Which is what I tried to say ;)

To put it a bit more concisely than my somewhat meandering previous comment:

we originally had (your code):

```bash
L=$(ls -dl /sbin/init 2>/dev/null)
```

my initial attempt to address the issue:

```bash
[ -f /sbin/init ] && L=$(ls -dl /sbin/init 2>/dev/null)
```

the check inside of the subshell – which did, unfortunately, not work, because it would still need something like `|| true` at the end:

```bash
L=$([ -f /sbin/init ] &&  ls -dl /sbin/init 2>/dev/null)
```

And now a potentially much simpler solution (just declaring `L` with `local` – which we can do, as we are in a function):

```bash
local L=$(ls -dl /sbin/init 2>/dev/null)
```

which seems to work fine in all shells and cases I checked (with/without `/sbin/init` existing).

Reply:
Not entirely – but almost. Which is what I tried to say ;)

To put it a bit more concisely than my somewhat meandering previous comment:

we originally had (your code):

```bash
L=$(ls -dl /sbin/init 2>/dev/null)
```

my initial attempt to address the issue:

```bash
[ -f /sbin/init ] && L=$(ls -dl /sbin/init 2>/dev/null)
```

the check inside of the subshell – which did, unfortunately, not work, because it would still need something like `|| true` at the end:

```bash
L=$([ -f /sbin/init ] &&  ls -dl /sbin/init 2>/dev/null)
```

And now a potentially much simpler solution (just declaring `L` with `local` – which we can do, as we are in a function):

```bash
local L=$(ls -dl /sbin/init 2>/dev/null)
```

which seems to work fine in all shells and cases I checked (with/without `/sbin/init` existing).

/edit

sorry, accidentally used a different/wrong github account for the comment – so posting it again.

Reply:
ah. to work in ksh, that requires a slight modification:
```sh
local L
L=$(ls -dl /sbin/init 2>/dev/null)
```

Reply:
The following only works in Bash because when command substitution (`$(command)`) is used to assign a value to a variable *while declaring it* (`local VAR`), error handling is suppressed for the command being substituted.

```bash
local L=$(ls -dl /sbin/init 2>/dev/null)
```
[See here](https://mywiki.wooledge.org/BashPitfalls#local_var.3D.24.28cmd.29) for more information, although there is no mention of what happens when `errexit` and/or `errtrace` (`set -e` and `set -E` respectively) are enabled.

I'm not sure whether to blame a Bash change or a `makepkg` update, but `nvm install` fails consistently on my Arch build server with any package that uses `nvm`, despite working previously with no change to this line.

Adding `set -x` before `nvm install` results in the following output because--according to `echo "$-"; trap`--`makepkg` invokes PKGBUILD functions with `set -e`, `set -E`, and an `ERR` trap that calls `exit 4`, which means that when `ls -dl` fails, `exit 4` terminates four nested subshells before the error is caught and `Binary download failed, trying source.` is printed.

```
+++++ ...
++++ HOST_ARCH=x86_64
++++ local NVM_ARCH
++++ case "${HOST_ARCH}" in
++++ NVM_ARCH=x64
+++++ ls -dl /sbin/init
++++++ error_function prepare
++++++ ((  ! BASH_SUBSHELL  ))
++++++ exit 4
++++ L=
+++++ error_function prepare
+++++ ((  ! BASH_SUBSHELL  ))
+++++ exit 4
+++ NVM_ARCH=
++++ error_function prepare
++++ ((  ! BASH_SUBSHELL  ))
++++ exit 4
++ SLUG=
+++ error_function prepare
+++ ((  ! BASH_SUBSHELL  ))
+++ exit 4
+ TARBALL=
+ ...
```

I assume the intention is for `nvm` to work in any shell environment regardless of its initial state, so adding `set +E` if `[[ $- == *E* ]]` to the existing check for `set -e` should resolve this while preventing similar failures in future.

Reply:
set -e should never be used anyways - but nvm has measures in place to ensure it’s not set for nvm itself. I’m not sure what the E option does?

Reply:
`set -e` is often used in non-interactive/CI contexts, but yes, I had noticed nvm re-invokes itself without it if needed (same for `set -a` and a non-standard `IFS`). 👍🏻 

`set -E` enables ERR trap inheritance in Bash. Without `-E`, setting an ERR trap like `trap 'exit 4' ERR` would only run `exit 4` if a command fails in the current scope, but with `-E`, `exit 4` would also be executed whenever a command fails in a function, command substitution or subshell invoked or created by the current scope or any of its children. The Bash manual puts it like this:

> **-E**&nbsp;&nbsp;&nbsp;If set, any trap on **ERR** is inherited by shell functions, command substitutions, and commands executed in a subshell environment.  The **ERR** trap is normally not inherited in such cases.

If an ERR trap exits non-zero (as it does when nvm is invoked by an Arch package build script), the effect of `-E` is identical to `-e`. Again, quoting the Bash manual:

> If a sigspec is **ERR**, the command arg is executed whenever a pipeline (which may consist of a single simple command), a list, or a compound command returns a non-zero exit status, subject to the following conditions.  The **ERR** trap is not executed if the failed command is part of the command list immediately following a **while** or **until** keyword, part of the test in an **if** statement, part of a command executed in a **&&** or **||** list except the command following the final **&&** or **||**, any command in a pipeline but the last, or if the command's return value is being inverted using **!**.  These  are  the same conditions obeyed by the **errexit** (**-e**) option.

I've prepared a fix that hopefully addresses all of this without side-effects on other shells and will open a PR soon.

Reply:
Thanks for getting to the bottom of that @lkrms, and for providing a much less hacky attempt to fix the underlying issue!

I guess we should close this PR then, as it's superseded/replaced by #2678?

Reply:
Let’s hold off; i may still land both.